pkg/sidecar-controller/snapshot_controller.go

```diff
@@ -93,73 +93,47 @@ func (ctrl *csiSnapshotSideCarController) syncContent(content *crdv1.VolumeSnaps
 func (ctrl *csiSnapshotSideCarController) deleteCSISnapshot(content *crdv1.VolumeSnapshotContent) error {
 	operationName := fmt.Sprintf("delete-%s", content.Name)
 	klog.V(5).Infof("schedule to delete snapshot, operation named %s", operationName)
-	ctrl.scheduleOperation(operationName, func() error {
-		return ctrl.deleteCSISnapshotOperation(content)
-	})
-	return nil
-}
-
-// scheduleOperation starts given asynchronous operation on given snapshot. It
-// makes sure there is no running operation with the same operationName
-func (ctrl *csiSnapshotSideCarController) scheduleOperation(operationName string, operation func() error) {
```

Comment:
Without removing the scheduleOperation, I'll have to add a createSnapshotWrapper to be called by both createSnapshotOperation and checkandUpdateContentStatusOperation.

Another problem with the scheduleOperation is that it won't return any error code.  Without the error code, we couldn't add a failed item back to the worker queue.  Retry will still happen with the resync period, but there was a review comment during the beta review to do the requeue.

Reply:
The scheduleOperation code is based on implementation of the in-tree PV/PVC controller.  I see that the in-tree PV/PVC controller also does not do requeue.   

The other in-tree controllers such as attach/detach controller, expand controller, and pv/pvc protection controller don't have scheduleOperation and all have the reuque logic.
The out-of-tree provisioner (https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) which is used together with external-provisioner also does not use scheduleOperation and it does have requeue logic.

Reply:
Xref: https://github.com/kubernetes/kubernetes/issues/85661

Imo, relying on a resync to requeue is not a good design pattern to continue. It then requires you have a frequent resync period to increase performance, which generates more unnecessary churn especially as the number of objects scale.

Reply:
ACK. That confirms that we should get rid of scheduleOperation so that we can properly do requeue.

Reply:
W.R.T that, I was proposing to use kubebuilder which provides fairly flexible requeueing mechanism (either by returning an error from Reconcile or specifying a polling period in return result), but that requires a total revamp of current implementation.
Now my question is do we want to have this big revamp in the current release given that there are other critical bug fixes.

Reply:
IMO, I'd have this enhancement in GA, and have an interim release in the middle for better testing.

Reply:
I don't want to use kubebuilder.  It hides lots of things.  Other controllers (not created by kubebuilder) work fine with the requeue logic.

Reply:
Also snapshot controller and csi-snapshotter sidecar are separate controllers.  Normally when you use kubebuilder, it would generate a controller that goes with the CRDs but that would be for the snapshot controller.  For the sidecar controller, we still need to write a separate one.

Reply:
true, and wasn't that the plan to separate sidecar/snapshot controller into two repos anyways?
again, I am not suggesting kubebuilder is the way to go :), its just an alternative could be considered.
my main concern here is really the potential racing I listed initially. can requeue happen inside of createSnapshotOperation? that would solve the problem of returning code?

Reply:
>  true, and wasn't that the plan to separate sidecar/snapshot controller into two repos anyways?

What I meant to say is that we may not be able to use kubebuilder for sidecar anyway.  The scheduleOperation code is only in the sidecar.  The common controller does not use scheduleOperation (needs to clean up unused code).

> my main concern here is really the potential racing I listed initially. can requeue happen inside of createSnapshotOperation? that would solve the problem of returning code?

We could look into that but that could make the code look very ugly.  Normally requeue happens in snapshot_controller_base.go.  If we do requeue inside that operation, we would need to do requeue in snapshot_controller.go.

How about this?  I'll remove the code that removes scheduleOperation from this PR and do that together with this other requeue PR later.  But that means I'll have to add a createSnapshotWrapper so that createSnapshotOperation won't be called by  checkandUpdateContentStatusOperation.

On another note, we probably don't need checkandUpdateContentStatusOperation any way.  Even in the in-tree PV/PVC controller, I only see operation being used for create/delete, but not on update.

Reply:
> > true, and wasn't that the plan to separate sidecar/snapshot controller into two repos anyways?
> 
> What I meant to say is that we may not be able to use kubebuilder for sidecar anyway. The scheduleOperation code is only in the sidecar. The common controller does not use scheduleOperation (needs to clean up unused code).
It's possible to use kubebuilder for sidecar as long as the CRD/API is in a separated repo.

> > my main concern here is really the potential racing I listed initially. can requeue happen inside of createSnapshotOperation? that would solve the problem of returning code?
> 
> We could look into that but that could make the code look very ugly. Normally requeue happens in snapshot_controller_base.go. If we do requeue inside that operation, we would need to do requeue in snapshot_controller.go.
you are right, that's not good. lets not moving towards that.

> How about this? I'll remove the code that removes scheduleOperation from this PR and do that together with this other requeue PR later. But that means I'll have to add a createSnapshotWrapper so that createSnapshotOperation won't be called by checkandUpdateContentStatusOperation.
SGTM, this PR is already complex, and I am a bit hesitate to include too many changes right before release cut which is an optimization not critical bug fix. WDYT?

> On another note, we probably don't need checkandUpdateContentStatusOperation any way. Even in the in-tree PV/PVC controller, I only see operation being used for create/delete, but not on update.
I complete agree. I think the current controller implementation could be improved by given more time.

Reply:
I'll break this PR into several PRs.
- This PR will only include the fix for create snapshot timeout.  This means I'll need to add that createSnapshotWrapper back:).  We want this fix for this release.
- I'll move the check for PVC is being used by the current snapshot to a separate PR as it is fixing the problem of removing the PVC finalizer.  We want this fix for this release.
- The changes related to scheduleOperation will be moved to a future PR as it is an optimization.